File: LeetCode/P155_MinStack.java
/*
 * LeetCode 155 - Min Stack
 * https://leetcode.com/problems/min-stack/
 *
 * Approach: Two stacks - the main stack, and a min-stack that tracks
 *           the minimum value at each level.
 * Time  : O(1) for all operations
 * Space : O(n)
 */

import java.util.Stack;

public class P155_MinStack {

    static class MinStack {
        private final Stack<Integer> stack;
        private final Stack<Integer> minStack;

        public MinStack() {
            stack = new Stack<>();
            minStack = new Stack<>();
        }

        public void push(int val) {
            stack.push(val);
            if (minStack.isEmpty() || val <= minStack.peek()) {
                minStack.push(val);
            } else {
                minStack.push(minStack.peek());
            }
        }

        public void pop() {
            stack.pop();
            minStack.pop();
        }

        public int top() {
            return stack.peek();
        }

        public int getMin() {
            return minStack.peek();
        }
    }

    // ---------- Simple manual test ----------
    public static void main(String[] args) {
        MinStack minStack = new MinStack();
        minStack.push(-2);
        minStack.push(0);
        minStack.push(-3);
        System.out.println(minStack.getMin()); // -3
        minStack.pop();
        System.out.println(minStack.top());    // 0
        System.out.println(minStack.getMin()); // -2
    }
}

File: LeetCode/P155_MinStack.cpp
#include <bits/stdc++.h>
using namespace std;

class MinStack {
private:
    stack<int> stk;
    stack<int> minStack;

public:
    MinStack() {
    }

    void push(int val) {
        stk.push(val);
        if (minStack.empty() || val <= minStack.top()) {
            minStack.push(val);
        } else {
            minStack.push(minStack.top());
        }
    }

    void pop() {
        stk.pop();
        minStack.pop();
    }

    int top() {
        return stk.top();
    }

    int getMin() {
        return minStack.top();
    }
};

int main() {
    MinStack minStack;
    minStack.push(-2);
    minStack.push(0);
    minStack.push(-3);
    cout << minStack.getMin() << "\n";
    minStack.pop();
    cout << minStack.top() << "\n";
    cout << minStack.getMin() << "\n";
    return 0;
}
